trio_asyncio/_loop.py

```diff
@@ -391,20 +391,45 @@ async def async_main(*args):
     """
 
     # TODO: make sure that there is no asyncio loop already running
-    async with trio.open_nursery() as nursery:
+
+    # The trio-asyncio loop can't shut down until all trio_as_aio tasks
+    # (or others using run_trio) have exited. This is because the
+    # termination of such a Trio task sets an asyncio future, which
+    # uses call_soon(), which won't work if the loop is closed.
+    # So, we use two nested nurseries.
+    async with trio.open_nursery() as loop_nursery:
         loop = TrioEventLoop(queue_len=queue_len)
         old_loop = current_loop.set(loop)
         try:
             loop._closed = False
-            await loop._main_loop_init(nursery)
-            await nursery.start(loop._main_loop)
-            yield loop
+            async with trio.open_nursery() as tasks_nursery:
+                await loop._main_loop_init(tasks_nursery)
+                await loop_nursery.start(loop._main_loop)
+                yield loop
+                tasks_nursery.cancel_scope.cancel()
+
+                # Allow all submitted run_trio() tasks calls a chance
+                # to start before the tasks_nursery closes, unless the
+                # loop stops (due to someone else calling stop())
+                # before that:
+                async with trio.open_nursery() as sync_nursery:
+                    sync_nursery.cancel_scope.shield = True
+
+                    @sync_nursery.start_soon
+                    async def wait_for_sync():
+                        try:
+                            await loop.synchronize()
+                        except RuntimeError:  # "Event loop is closed"
+                            pass
```

Comment:
Looks a bit dangerous to ignore every `RuntimeError` here. I've made a PR (against this branch) with a possible improvement :) 

https://github.com/oremanj/trio-asyncio/pull/1

Reply:
I think it's the lesser of multiple evils, since it's the exception that asyncio raises in this case and we want to mimic their API; and the operation is simple enough that there aren't other plausible sources of RuntimeError. Moreover, if something is screwy and this raises RuntimeError for a different reason, I contend that skipping this synchronization step (and accepting whatever error ensues) is better than deadlocking (which is the likely other outcome). Your PR doesn't export the `EventLoopClosedError` exception, so it's possible for it to create user confusion if it shows up in a traceback but they can't catch it.

IOW: it would've been nice if asyncio used a custom exception for this condition, but given that they don't, I think it's better for us not to add our own.

Reply:
Why not
```
try:
    loop._check_closed()
except RuntimeError:
    pass
else:
    await loop.synchronize()
```
?
That'd do this check twice, but that's too cheap to be noticeable IMHO.

Reply:
That seems fine to me. I didn't go that route because I didn't want to rely on a private member, but we already do elsewhere, so...

Reply:
> I think it's the lesser of multiple evils

Agreed!

Reply:
@oremanj I like @smurfix's idea to call `_check_closed()` beforehand.

(As for it being a private member - in a separate PR we can switch to using public [`is_closed()`][1] method everywhere. It was introduced in Python 3.5.)

[1]: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.is_closed

Reply:
`is_closed()` is perfect here, I didn't realize it existed! Then we don't have to worry about the exception at all.